### PR TITLE
make `take` behave like `first`

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description

Fixes: #6890 
But I have a question: why `take` command exists, it seems that it does nearly the same to `first`.

# Tests

Make sure you've done the following:

- [X] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [X] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [X] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [X] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [X] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [X] `cargo test --workspace --features=extra` to check that all the tests pass

# Documentation

- [N/A] If your PR touches a user-facing nushell feature then make sure that there is an entry in the documentation (https://github.com/nushell/nushell.github.io) for the feature, and update it if necessary.

### CODE REVIEW

**Comment by sholderbach** [comment]
Something in me revolts against landing this change just now. 

1. It can break peoples code (so changes in scripts and documentation plus release note entry for the breaking change)
2. We have the similar first command. Should this become an alias, be removed, or is there some useful semantics that should be different from the `first`/`last` pair. `take` implies to me some iterator or given order but we do not have this as a first class concept but it is rather an implementation detail.

On the previous note but slightly unrelated, there is some potential unintended ambiguity for `get 1` in the table context. Would it be helpful to have row/major axis indexing command that reduces the rank (i.e. strips the dimension like `get`)? (I think `take` would indeed not be the best fit thinking from `.take(3)` in rust land)

**Comment by fdncred** [comment]
IMO, I think we could change `first`/`last`/`take` in a few ways. I'm not sure which is right.

1. leave them alone and just allow there to be separate ways to do the same thing
2. make `first` only retrieve the very first item and not take parameters
3. deprecate `first` in favor of `take`
4. make `last` only retrieve the very last item and not take parameters 
5. if we do number 4 we should make `take` take negative parameters in order to replicate `last` functionality

one of the big differences previously was that `take` would work on binary and `first` wouldn't. I'm not sure it still exists but that was one of the reasons it was created IIRC.

**Comment by WindSoilder** [comment]
I vote for 2,3,4,5.

> one of the big differences previously was that take would work on binary and first wouldn't. I'm not sure it still exists but that was one of the reasons it was created IIRC.

Just check the implementation, `take` and `first` both work on binary

**Comment by webbedspace** [comment]
>leave them alone and just allow there to be separate ways to do the same thing

>deprecate first in favor of take

I kind of assumed these two existed together because they needed to be antonyms of `drop` and `last`, respectively. Though, I think the `take`/`drop` symmetry is much weaker than the `first`/`last` symmetry. Also, I personally think `take` sounds too similar to `get`. Thus, I think `take` should be deprecated in favour of `first`, **not** vice-versa. (Although, if `take` has a much longer legacy, that may complicate matters...)

**Comment by rgwood** [comment]
> 2. make first only retrieve the very first item and not take parameters

i.e.
|  Command | Return Type  | 
|---|---|
|  `first` |  `T` |  
| `take` |  `List<T>` |  

I think I like this best. It's consistent with LINQ from .NET and the popular `lodash` JS library.

> `take 3` sounds less descriptive of what it actually does

I think `take` is fairly common elsewhere ([.NET](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.take?view=net-7.0), [JS](https://lodash.com/docs/4.17.15#take), [Haskell](http://zvon.org/other/haskell/Outputprelude/take_f.html)) and so some users will already expect the name. But I can see how `first` is more natural in English.

**Comment by fdncred** [comment]
Good points @rgwood. Yes, let's do it that way please.

**Comment by webbedspace** [comment]
Personally I don't like the idea of `first` only taking the first value because then it's literally just `get 0`, even with the same character length. At that point the only reason to keep it is for symmetry with `last`, which at least is a bit more useful.

**Comment by sholderbach** [comment]
@webbedspace I think you have at the moment the best overview about the consistencies or inconsistencies. 

How about we summarize the current state in a matrix how each command that does indexing as a (side) function behaves under different inputs:
- On what container types does it operate
- Rows/Columns on tables
- Return types.
- Ranging selection vs individual entries
- how it deals with missing data/errors

If we have an overview we can probably better see what is weird or which common usage has no nice idiomatic approach.
I think only after that it makes sense to spend time bikeshedding favorite command names.

**Comment by dandavison** [comment]
I agree with that this is a good area to be working on, but I actually don't think that a PR is the right next step here. Here we're having design discussions in a PR, and we should avoid that: for user-facing changes, if we're still having design discussions, then ideally we shouldn't be writing any code. Instead, we could have the design discussions in issues (like the several excellent issues @webbedspace has opened). Or we could create a `proposals/` folder and open PRs with markdown proposals and no code. (Maybe the core team would like to give direction on how we should do design proposals for user-facing changes?)

@sholderbach it seems that you and me were thinking very similar things in the last few minutes!

@WindSoilder would you be interested in finishing this table off and turning it into a proposal?

# Problem

Commands such as `first`, `last`, `take`, `skip` have overlapping functionality, and variable output types depending on positional arguments. In some places they differ from libraries in other programming languages in surprising ways. Recently there have been several discussions about how we should design this area; it's clear that most nushell designers aren't happy with all aspects the current design.

# Background

The following table attempts to summarize the situation:

|                               |                   |                  |                  |                      |                       |                             |
| ----------------------------- | ----------------- | ---------------- | ---------------- | -------------------- | --------------------- | --------------------------- |
| **Functionality**             | **First element** | **n-th element** | **Last element** | **First n elements** | **Slice**             | **Slice from n-th element** |
| **Return Type**               | `T`               | `T`              | `T`              | `List<T>`            | `List<T>`             | `List<T>`                   |
| **Python** `(more_)itertools` | `first `          | `nth`            | `last`           | `take(n)`            | `islice(start, stop)` | `islice(start, None)`       |
| **Javascript** (`lodash`)     |                   |                  |                  |                      |                       |                             |
| **Rust**                      | `nth(0)`          | `nth()`          |                  |                      |                       |                             |
| **Scala**                     | `head`            | `drop(n-1).next` | `last`           | `take(n)`            | `slice(start, stop)`  | `drop(n)`                   |
| **More languages...**         |                   |                  |                  |                      |                       |                             |
| **Nushell (current)**         | `first`, `take`   | Does not exist   | `last`           | `first n`, `take n`  | Does not exist        | `skip n`                    |
| **Nushell (proposed)**        | ???               | ???              | ???              | ???                  | ???                   | ???                         |


```
|                               |                   |                  |                  |                      |                       |                             |
| ----------------------------- | ----------------- | ---------------- | ---------------- | -------------------- | --------------------- | --------------------------- |
| **Functionality**             | **First element** | **n-th element** | **Last element** | **First n elements** | **Slice**             | **Slice from n-th element** |
| **Return Type**               | `T`               | `T`              | `T`              | `List<T>`            | `List<T>`             | `List<T>`                   |
| **Python** `(more_)itertools` | `first `          | `nth`            | `last`           | `take(n)`            | `islice(start, stop)` | `islice(start, None)`       |
| **Javascript** (`lodash`)     |                   |                  |                  |                      |                       |                             |
| **Rust**                      | `nth(0)`          | `nth()`          |                  |                      |                       |                             |
| **Scala**                     | `head`            | `drop(n-1).next` | `last`           | `take(n)`            | `slice(start, stop)`  | `drop(n)`                   |
| **More languages...**         |                   |                  |                  |                      |                       |                             |
| **Nushell (current)**         | `first`, `take`   | Does not exist   | `last`           | `first n`, `take n`  | Does not exist        | `skip n`                    |
| **Nushell (proposed)**        | ???               | ???              | ???              | ???                  | ???                   | ???                         |
```

**Comment by fdncred** [comment]
@dandavison we do have https://github.com/nushell/rfcs and have discussed internally about maybe using it again. seems like it may fit what you're describing. thoughts?

**Comment by dandavison** [comment]
> we do have https://github.com/nushell/rfcs 

Thanks, that looks like exactly what I'm talking about.

**Comment by kubouch** [comment]
@dandavison n-th element in Nushell is `get` and slice is `range`.

**Comment by dandavison** [comment]
> n-th element in Nushell is `get` and slice is `range`.

Thank you @kubouch! Updated.

(I had been thinking of `get` as for key/column lookups on records/tables, and had not internalized that it does array/row indexing on lists/tables. And `range` I wasn't aware of; I had just searched for `slice`, and the terms aren't linked by help search terms. Most languages (e.g. Python, Rust, Scala, R, JS lodash) use "range" to refer to a constructor yielding some sort of sequence, so nushell should probably follow suit and rename range to slice unless there was a reason to differ.)

**Comment by webbedspace** [comment]
> @webbedspace I think you have at the moment the best overview about the consistencies or inconsistencies.
> 
> How about we summarize the current state in a matrix how each command that does indexing as a (side) function behaves under different inputs:

Actually, I've been working on something like this myself in private, but covering the entire basic (non-dataframe non-database) data structure command suite: https://docs.google.com/spreadsheets/d/1Hd6rdMKH1a-ttLsilvFZ0x_Mya7e3aGYBDtSEC6piDI/ This is based entirely on my observations and any inaccuracies are welcomed.

**Comment by dandavison** [comment]
> Actually, I've been working on something like this myself in private, but covering the entire basic (non-dataframe non-database) data structure command suite: https://docs.google.com/spreadsheets/d/1Hd6rdMKH1a-ttLsilvFZ0x_Mya7e3aGYBDtSEC6piDI/ 

This looks great! Do you want to open up commenting permissions? Honestly, I think Google sheets/docs or other collaborative editing tools might be better for collaborating on documentation and design proposals than markdown PRs.

**Comment by webbedspace** [comment]
Some big things I haven't covered in that sheet yet:
* The flags situation (kind of complicated)
* Fully disambiguating how many indexes/names can be given to something that wants index I or name A
* Cell paths. For instance, `[{a:{a:[1 2]}}] | get 0.a.a.1` works but `[{a:{a:[1 2]}}] | select 0.a.a.1` (justifiably) doesn't. However, `[{a:{a:[1 2]}}] | reject 0.a.a.1` also works, which is desirable but complicates the name symmetry.

**Comment by WindSoilder** [comment]
Since @webbedspace have provide a good doc about overview of nushell commands.

For this pr, I think it's ok to take a step back and see how we will come up with a decision about command consistencies.

**Comment by fdncred** [comment]
team, are we ready for review here - or is the discussion still happening?

**Comment by dandavison** [comment]
I think the change is correct (`take` should always return a list; `take 1` shouldn't magically unwrap the result). But seeing as there are other breaking changes we will probably make to related commands, I think we should close this PR for now, and open a proposal in the proposals repo based on the tables and google sheets above, covering all proposed naming and behavior changes to commands such as `first`,`last`,`take`,`skip`,`reject`,`drop`,`slice`.

**Comment by dandavison** [comment]
As yet another random user disagreement... :) I disagree with the behaviour

```
> [1 2 3] | take
1
```

IMO this should be an error; in other languages the API is `take n` with an argument.

I'm just mentioning this to emphasize how many disagreements people have with the current design and the need for a design-first instead of code-first approach to the nushell builtin commands collection.

**Comment by rgwood** [comment]
I'd be open to merging this instead of waiting for a larger change, but I don't feel strongly about it.

Otherwise, I very much agree with Dan.

**Comment by fdncred** [comment]
ya, I'd vote to land and then follow the rest of Dan's suggestion.

File: crates/nu-command/src/filters/take/take_.rs [hunk 1]
```diff
@@ -138,7 +146,7 @@ fn first_helper(
                 }
             }
             _ => {
-                if rows_desired == 1 {
+                if rows_desired == 1 && rows.is_none() {
```
Comment (dandavison) [inline crates/nu-command/src/filters/take/take_.rs:149]: @rgwood @fdncred Right, fair enough, I agree merging makes sense. I don't actually agree with this line, since I don't think we should ever do the `rows_desired == 1` behaviour, but this PR makes things better not worse than the current state.

Comment (WindSoilder) [reply crates/nu-command/src/filters/take/take_.rs:149]: Ah, I just copy the code from `first` command, I may need to take a look in detail :-)

**Comment by kubouch** [comment]
Seems like we all agree that `take` should always take a number, that is, `ls | take` should error. If you see `ls | take` in a script you wouldn't know what it's supposed to do (unlike `ls | first` which IMO justifies the existence of `first`).

I'd say, @WindSoilder you can change `take` to always require a parameter, then we'll land it, then we can have a broader discussion about consistency of the filtering commands.

**Comment by kubouch** [comment]
OK, cool. Could you also add a test for the `... | take` error so we can track it in the future?

**Comment by kubouch** [comment]
Thanks! I think it's still good to fix the inconsistency. We just had a release and can revise this in subsequent PRs if necessary.

**Comment by fdncred** [comment]
BTW - the title of this pr no longer matches this functionality - take doesn't behave like first - LOL
Screenshot 2022-11-09 at 6 08 11 AM
